### PR TITLE
added margin_ranking_loss

### LINKED ISSUE #21074: margin_ranking_loss

### PR DESCRIPTION
**Pull request by sushmanthreddy**
close #21074

### CODE REVIEW

**Comment by tomatillos** [comment]
Hi, great start but the tests are not passing - would you mind taking a look? Let me know if you need any help.

**Comment by sushmanthreddy** [comment]
@tomatillos can u plss tell me , how to run the tests?
i have been stuck

**Comment by tomatillos** [comment]
Take a look at: https://unify.ai/docs/ivy/overview/contributing/setting_up.html, and find the relevant section to which setup you are using.

**Comment by tomatillos** [comment]
Good work so far. For this subtask, you need to implement the function in all of the backends, not just tensorflow. Let me know if you need any help.

**Comment by sushmanthreddy** [comment]
in which file should i add backend support?
in backend, there is no support file regarding losses

**Comment by tomatillos** [comment]
Actually, the task description says to implement the loss function in experimental (see https://github.com/unifyai/ivy/issues/3856). This will fix the problem with the tests. Also you should get the lint test passing, you can setup the pre-commit here: https://unify.ai/docs/ivy/overview/contributing/setting_up.html#pre-commit

**Comment by sushmanthreddy** [comment]
> Actually, the task description says to implement the loss function in experimental (see #3856). This will fix the problem with the tests. Also you should get the lint test passing, you can setup the pre-commit here: https://unify.ai/docs/ivy/overview/contributing/setting_up.html#pre-commit

thank you..i will keep new pr soon for lint passing

**Comment by sushmanthreddy** [comment]
@tomatillos the code has been shifted to the experimental folder ... can u pls review it and merge it

**Comment by tomatillos** [comment]
Hi thanks, do you mind getting the lint passing, and removing the commits from the file you're not adding anything to (ivy/functional/ivy/losses.py). Then we can get this merged! :)

**Comment by sushmanthreddy** [comment]
@tomatillos plss it has passed the link testing plss merge it

**Comment by sushmanthreddy** [comment]
@tomatillos docstrings are updated

**Comment by tomatillos** [comment]
Hi, the tests are not passing on the CI, do you have the testing environment setup locally?

**Comment by sushmanthreddy** [comment]
@tomatillos The margin_ranking_loss support for the ivy/experimental is added here.The test cases are passing here for this
Screenshot from 2023-08-17 21-59-30

The stateful test cases are failing because of the ivy pipeline is for the stateful is not complete yet and paddle bfloat16 issues is still pending for ivy lib.

As these margin_ranking_loss is mixed function backend implementation is only there for paddle and torch.so i have added them in the backen loss

File: ivy/data_classes/container/experimental/losses.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        out
+            optional output container, for writing the result to. It must have a shape
+            that the inputs broadcast to.
+
+        Returns
+        -------
+        ret
+            The margin ranking loss.
+
+        Examples
```
Comment (tomatillos) [inline ivy/data_classes/container/experimental/losses.py:144]: What is going on in this example?

Comment (sushmanthreddy) [reply ivy/data_classes/container/experimental/losses.py:144]: @tomatillos The code provides examples of how to use the methods defined in the class. In both examples, they show how to create input Container instances, where each container has multiple named sub-containers. These sub-containers seem to correspond to different data samples or data points. The examples then demonstrate how to calculate the margin ranking loss for the provided inputs.

**Comment by sushmanthreddy** [comment]
@tomatillos I have mentioned the changes ,can u pls review it ?and merge it too...if there are no mistakes?

**Comment by sushmanthreddy** [comment]
@tomatillos checked the datatypes by this method and added it.and cross verfied with documentation also.
Screenshot from 2023-08-20 02-35-57

File: ivy_tests/array_api_testing/test_array_api
Comment (tomatillos) [inline ivy_tests/array_api_testing/test_array_api:1]: Can we remove this from the PR?

**Comment by sushmanthreddy** [comment]
@i have gone through line to line and changed everything , now tests are passing doc strings are correct and values in them also are correct .
I have added supported dtypes and unsupported using above mentioned strategy and removed test_array_api as u mentioned ,u have mentioned that in my other pr that tests are failing .It was never because of my written code ,it was due to ivy  testing pipeline .Here is the proof of it [link](https://discord.com/channels/799879767196958751/982738436383445073/1143422045271629894)

And sorry if have made u feel rush .

I was just excited in merging my code into ivy code base .Now tests are passing except for the paddle due to bfloat16 issue .

Screenshot from 2023-08-22 22-44-55
here ,I have updated my code by using git pull and tested changes in my local machine .
Screenshot from 2023-08-22 22-45-05
u can see only now tests are passing except for paddle and that is due to bfloat16 issue 
Screenshot from 2023-08-22 22-50-07


As now tests are  passing .I don't knew ivy CI is still mess ,its causing testing issues.
I hope that this above screenshots will help u out for checks passing and documentation is made according to your past reviews
